src/main/java/seedu/address/model/tag/UniqueTagList.java

```diff
@@ -0,0 +1,139 @@
+import javafx.collections.ObservableList;
+import seedu.address.model.person.exceptions.DuplicatePersonException;
+import seedu.address.model.tag.exceptions.DuplicateTagException;
+import seedu.address.model.tag.exceptions.TagNotFoundException;
+
+/**
+ * A list of tags that enforces uniqueness between its elements and does not allow nulls.
+ * A tag is considered unique by comparing using {@code Tag#isSameTag(Tag)}. As such, adding and updating of
+ * tags uses Tag#isSameTag(Tag) for equality so as to ensure that the person being added or updated is
+ * unique in terms of identity in the UniqueTagList. However, the removal of a tag uses Tag#equals(Object) so
+ * as to ensure that the tag with exactly the same fields will be removed.
+ *
+ * Supports a minimal set of list operations.
+ *
+ * @see Tag#isSameTag(Tag)
+ */
+public class UniqueTagList implements Iterable<Tag> {
+    private final ObservableList<Tag> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Tag> internalUnmodifiableList =
+            FXCollections.unmodifiableObservableList(internalList);
+
+    /**
+     * Returns true if the list contains an equivalent tag as the given argument.
+     */
+    public boolean contains(Tag toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::isSameTag);
+    }
+
+    /**
+     * Adds a tag  to the list.
+     * The tag must not already exist in the list.
+     */
+    public void add(Tag toAdd) {
+        requireNonNull(toAdd);
+        if (contains(toAdd)) {
+            throw new DuplicateTagException();
+        }
+        internalList.add(toAdd);
+    }
+
+    /**
+     * Replaces the tag {@code target} in the list with {@code editedTag}.
+     * {@code target} must exist in the list.
+     * The tag of {@code editedTag} must not be the same as another existing tag in the list.
+     */
+    public void setTag(Tag target, Tag editedTag) {
+        requireAllNonNull(target, editedTag);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new TagNotFoundException();
+        }
+
+        if (!target.isSameTag(editedTag) && contains(editedTag)) {
+            throw new DuplicateTagException();
+        }
+
+        internalList.set(index, editedTag);
+    }
+
+    public void setTags(UniqueTagList replacement) {
+        requireNonNull(replacement);
+        internalList.setAll(replacement.internalList);
+    }
+
+    /**
+     * Replaces the contents of this list with {@code tags}.
+     * {@code tags} must not contain duplicate tags.
+     */
+    public void setTags(List<Tag> tags) {
+        requireAllNonNull(tags);
+        if (!tagsAreUnique(tags)) {
+            throw new DuplicatePersonException();
+        }
+
+        internalList.setAll(tags);
+    }
+
+    /**
+     * Returns true if {@code tags} contains only unique tags.
+     */
+    private boolean tagsAreUnique(List<Tag> tags) {
```

Comment:
naming of this method can be improved as `areTagsUnique` to match coding standard given.